### PR TITLE
Eliminate a source of code bloat in the kani library

### LINKED ISSUE #2010: Warning about unsupported `caller_location` intrinsic
Kani almost always emits the following annoying warning on any program:
```
warning: Found the following unsupported constructs:
             - caller_location (1)
         
         Verification will fail if one or more of these constructs is reachable.
         See https://model-checking.github.io/kani/rust-feature-support.html for more details.

warning: 1 warning emitted
```
for example, on this tiny program:

```rust
#[kani::proof]
fn main() {
    let _x: bool = kani::any();
}
```

using the following command line invocation:

```
kani test.rs
```

with Kani version: f4796eacbdae07a4cbd305bebabbc2364f8bddd9

It's not clear what the source of this intrinsic is, so it might be worth investigating if this is coming from codegen itself.

### PR DESCRIPTION
**Pull request by zhassan-aws**
### Description of changes: 

Same trick applied in #2032, but this time to the `kani::panic` function. In addition to eliminating the `caller_location` warning on a small program like this one:
```rust
#[kani::proof]
fn main() {
    panic!("Some message");
}
```
it also results in a speedup and a significant reduction in the MIR size.

Before:
```bash
$ /usr/bin/time -p kani simp_panic.rs
...
real 3.92
user 3.54
sys 0.37

$ RUSTFLAGS="--emit mir" kani simp_panic.rs
$ wc simp_panic.kani.mir 
  33248  237545 4173632 simp_panic.kani.mir
```
After:
```bash
$ /usr/bin/time -p kani simp_panic.rs
...
real 0.15
user 0.11
sys 0.04

$ RUSTFLAGS="--emit mir" kani simp_panic.rs
$ wc simp_panic.kani.mir 
  86  530 5399 simp_panic.kani.mir
```

### Resolved issues:

Resolves #2010 

### Related RFC:

Optional #ISSUE-NUMBER.

### Testing:

* How is this change tested? Couldn't find a way to test this.

* Is this a refactor change? No

### Checklist
- [X] Each commit message has a non-empty body, explaining why the change was made
- [X] Methods or procedures are documented
- [X] Regression or unit tests are included, or existing tests cover the modified code
- [X] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: library/kani/src/lib.rs [hunk 1]
```diff
@@ -154,8 +148,22 @@ fn any_raw_inner<T>() -> T {
...
+/// This is a dummy function that is intended to be called from functions with
+/// `rustc_diagnostic_item`, i.e. that have kani hooks, to make sure they
+/// compile.
+/// While we could use `unreachable!()` or `panic!()` as the body of this
+/// function, both cause Kani to produce a warning on missing `caller_location`
+/// intrinsic (see https://github.com/model-checking/kani/issues/2010) and
+/// produce unnecessarily large MIR.  This function is handled via a hook
+/// anyway, so we just need to put a body that rustc does not complain about. An
+/// infinite loop works out nicely.
+const fn __kani_dummy_never() -> ! {
```
Comment (tedinski) [inline library/kani/src/lib.rs:164]: `#[doc(hidden)]` for things we don't want customer-facing.

**Review by tedinski — APPROVED** [review]
I'm slightly concerned about playing whack-a-mole with this.

With assess, I think we also see a lot of code bloat and caller_location. I haven't looked into it yet, but it'd be nice if we could integrate a hook or something in Kani that fixed all this everywhere, and then we could revert these back to being basic `unreachable!()` or whatever.

File: library/kani/src/lib.rs (hunk 1 above)
Comment (zhassan-aws) [reply library/kani/src/lib.rs:164]: Is this needed for a private function?

Comment (tedinski) [reply library/kani/src/lib.rs:164]: Oh, good point. Nevermind then :)

**Comment by celinval** [comment]
I don't think hooks will help here, since it would run after the reachability analysis. We could potentially plug hooks into the reachability analysis or use stubs here.

For the caller location warning, we should probably just fix it.

**Comment by celinval** [comment]
Also, I mentioned this to @zhassan-aws. Fixing our library is not really going to fix the problem in real user code, since any call to the std library that may eventually panic will bring in all that code. E.g.: `unwrap()`.

**Comment by zhassan-aws** [comment]
> Fixing our library is not really going to fix the problem in real user code, since any call to the std library that may eventually panic will bring in all that code. E.g.: unwrap().

True. One interesting thing though is that `panic` in particular seems to be excessively causing a MIR code bloat:
```
$ cat simp_panic.rs 
#[kani::proof]
fn main() {
    panic!("Some message");
}
$ RUSTFLAGS="--emit mir" kani simp_panic.rs
$ wc simp_panic.kani.mir 
  33248  237545 4173632 simp_panic.kani.mir
```

```
$ cat simp_unwrap.rs
#[kani::proof]
fn main() {
    let x: Option<i32> = kani::any();
    x.unwrap();
}
$ RUSTFLAGS="--emit mir" kani simp_unwrap.rs
$ wc simp_unwrap.kani.mir 
  1083   7418 124029 simp_unwrap.kani.mir
```
I'm not sure what else `panic` brings other than formatting code (which `unwrap` also bring as @celinval pointed out) that is causing the number of lines of MIR to be 33X that of the unwrap program.

These numbers are using c135fa9017f.

**Comment by zhassan-aws** [comment]
I evaluated the impact of this change on a few customer tests and found the reduction in the MIR size to be minimal. Abandoning this PR.

**Pull request closed without merging**